src/sign/records.rs

```diff
@@ -74,34 +77,122 @@ impl<N, D> SortedRecords<N, D> {
         self.rrsets().find(|rrset| rrset.rtype() == Rtype::SOA)
     }
 
+    /// Sign a zone using the given keys.
+    ///
+    /// A DNSKEY RR will be output for each key.
+    ///
+    /// Keys with a supported algorithm with the ZONE flag set will be used as
+    /// ZSKs.
+    ///
+    /// Keys with a supported algorithm with the ZONE flag AND the SEP flag
+    /// set will be used as KSKs.
+    ///
+    /// If only one key has a supported algorithm and has the ZONE flag set
+    /// AND has the SEP flag set, it will be used as a CSK (i.e. both KSK and
+    /// ZSK).
     #[allow(clippy::type_complexity)]
-    pub fn sign<Octets, Key, ApexName>(
+    pub fn sign<Octets, SigningKey>(
         &self,
-        apex: &FamilyName<ApexName>,
+        apex: &FamilyName<N>,
         expiration: Timestamp,
         inception: Timestamp,
-        key: Key,
-    ) -> Result<Vec<Record<N, Rrsig<Octets, ApexName>>>, Key::Error>
+        keys: &[(SigningKey, validate::Key<Octets>)], // private, public key pair
+    ) -> Result<Vec<Record<N, ZoneRecordData<Octets, N>>>, ()>
     where
         N: ToName + Clone,
-        D: RecordData + ComposeRecordData,
-        Key: SigningKey,
-        Octets: From<Key::Signature> + AsRef<[u8]>,
-        ApexName: ToName + Clone,
+        D: RecordData + ComposeRecordData + From<Dnskey<Octets>>,
+        SigningKey: SignRaw,
+        Octets: AsRef<[u8]>
+            + Clone
+            + From<Box<[u8]>>
+            + octseq::OctetsFrom<std::vec::Vec<u8>>,
     {
-        let mut res = Vec::new();
-        let mut buf = Vec::new();
+        // Per RFC 8624 section 3.1 "DNSSEC Signing" column guidance.
+        let unsupported_algorithms = [
+            SecAlg::RSAMD5,
+            SecAlg::DSA,
+            SecAlg::DSA_NSEC3_SHA1,
+            SecAlg::ECC_GOST,
+        ];
```

Comment:
I don't think this step is necessary.  Keys with must-not-implement algorithms aren't supported by `SigningKey` anyway.  It is the user's responsibility to select appropriate keys and to be aware of algorithm security.